// angular-baidu-maps/index.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { AbmComponent } from './core/abm.component';
import { AbmPanoramaComponent } from './core/abm-panorama.component';
import { AbmConfig } from './core/abm.config';
import { LoaderService } from './core/loader.service';
export { AbmConfig } from './core/abm.config';
export { AbmComponent } from './core/abm.component';
export { AbmPanoramaComponent } from './core/abm-panorama.component';
var AbmModule = (function () {
    function AbmModule() {
    }
    AbmModule.forRoot = function (config) {
        return {
            ngModule: AbmModule,
            providers: [
                { provide: AbmConfig, useValue: config }
            ]
        };
    };
    AbmModule.decorators = [
        { type: NgModule, args: [{
                    imports: [CommonModule],
                    providers: [LoaderService],
                    declarations: [AbmComponent, AbmPanoramaComponent],
                    exports: [AbmComponent, AbmPanoramaComponent]
                },] },
    ];
    /** @nocollapse */
    AbmModule.ctorParameters = function () { return []; };
    return AbmModule;
}());
export { AbmModule };
