// @delon/abc/utils/utils.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { LazyService } from './lazy.service';
// import { TrendComponent } from './trend.component';
var COMPONENTS = [];
// region: zorro modules
// import { NzBreadCrumbModule } from 'ng-zorro-antd';
// import { NzIconModule } from 'ng-zorro-antd-extra';
var ZORROMODULES = [];
// endregion
var AdUtilsModule = /** @class */ (function () {
    function AdUtilsModule() {
    }
    AdUtilsModule.forRoot = function () {
        return {
            ngModule: AdUtilsModule,
            providers: [
                LazyService
            ]
        };
    };
    AdUtilsModule.decorators = [
        { type: NgModule, args: [{
                    imports: [CommonModule].concat(ZORROMODULES),
                    declarations: COMPONENTS.slice(),
                    exports: COMPONENTS.slice()
                },] },
    ];
    /** @nocollapse */
    AdUtilsModule.ctorParameters = function () { return []; };
    return AdUtilsModule;
}());
export { AdUtilsModule };
