// @delon/abc/full-content/full-content.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FullContentComponent } from './full-content.component';
import { FullContentToggleDirective } from './full-content-toggle.directive';
import { FullContentService } from './full-content.service';
var COMPONENTS = [FullContentComponent, FullContentToggleDirective];
var AdFullContentModule = /** @class */ (function () {
    function AdFullContentModule() {
    }
    AdFullContentModule.forRoot = function () {
        return {
            ngModule: AdFullContentModule,
            providers: [
                FullContentService
            ]
        };
    };
    AdFullContentModule.decorators = [
        { type: NgModule, args: [{
                    imports: [CommonModule],
                    declarations: COMPONENTS.slice(),
                    exports: COMPONENTS.slice()
                },] },
    ];
    /** @nocollapse */
    AdFullContentModule.ctorParameters = function () { return []; };
    return AdFullContentModule;
}());
export { AdFullContentModule };
