// ng-zorro-antd-extra/nz-divider/nz-divider.component.js
import { Component, ViewEncapsulation, TemplateRef, Input, ElementRef, Renderer2 } from '@angular/core';
var NzDividerComponent = (function () {
    function NzDividerComponent(el, renderer) {
        this.el = el;
        this.renderer = renderer;
        // region fields
        this.title = false;
        this._title = '';
        this.nzType = 'horizontal';
        this.nzDashed = false;
        // endregion
        this._classMap = [];
    }
    Object.defineProperty(NzDividerComponent.prototype, "nzTitle", {
        set: function (value) {
            if (value instanceof TemplateRef)
                this._titleTpl = value;
            else
                this._title = value;
            this.title = !!value;
        },
        enumerable: true,
        configurable: true
    });
    NzDividerComponent.prototype.setClass = function () {
        var _this = this;
        this._classMap.forEach(function (cls) { return _this.renderer.removeClass(_this.el.nativeElement, cls); });
        this._classMap = ['ant-divider', "ant-divider-" + this.nzType];
        if (this.title)
            this._classMap.push("ant-divider-with-text");
        if (typeof this.nzDashed !== 'undefined' && this.nzDashed !== false)
            this._classMap.push("ant-divider-dashed");
        this._classMap.forEach(function (cls) { return _this.renderer.addClass(_this.el.nativeElement, cls); });
    };
    NzDividerComponent.prototype.ngOnChanges = function (changes) {
        this.setClass();
    };
    NzDividerComponent.prototype.ngOnInit = function () {
        this.setClass();
    };
    NzDividerComponent.decorators = [
        { type: Component, args: [{
                    selector: 'nz-divider',
                    template: "\n    <span *ngIf=\"title\" class=\"ant-divider-inner-text\">\n        <ng-container *ngIf=\"_title; else _titleTpl\">{{ _title }}</ng-container>\n    </span>\n    ",
                    styles: ["/* stylelint-disable at-rule-empty-line-before,at-rule-name-space-after,at-rule-no-unknown */ /* stylelint-disable declaration-bang-space-before */ /* stylelint-disable declaration-bang-space-before */ .ant-divider { font-family: \"Helvetica Neue For Number\", -apple-system, BlinkMacSystemFont, \"Segoe UI\", Roboto, \"PingFang SC\", \"Hiragino Sans GB\", \"Microsoft YaHei\", \"Helvetica Neue\", Helvetica, Arial, sans-serif; font-size: 14px; line-height: 1.5; color: rgba(0, 0, 0, 0.65); box-sizing: border-box; margin: 0; padding: 0; list-style: none; background: #e8e8e8; } .ant-divider, .ant-divider-vertical { margin: 0 8px; display: inline-block; height: 0.9em; width: 1px; vertical-align: middle; position: relative; top: -0.06em; } .ant-divider-horizontal { display: block; height: 1px; width: 100%; margin: 24px 0; } .ant-divider-horizontal.ant-divider-with-text { display: table; white-space: nowrap; text-align: center; background: transparent; font-weight: 500; color: rgba(0, 0, 0, 0.85); font-size: 16px; margin: 16px 0; } .ant-divider-horizontal.ant-divider-with-text:before, .ant-divider-horizontal.ant-divider-with-text:after { content: ''; display: table-cell; position: relative; top: 50%; width: 50%; border-top: 1px solid #e8e8e8; transform: translateY(50%); } .ant-divider-inner-text { display: inline-block; padding: 0 24px; } .ant-divider-dashed { background: none; border-top: 1px dashed #e8e8e8; } .ant-divider-horizontal.ant-divider-with-text.ant-divider-dashed { border-top: 0; } .ant-divider-horizontal.ant-divider-with-text.ant-divider-dashed:before, .ant-divider-horizontal.ant-divider-with-text.ant-divider-dashed:after { border-style: dashed none none; } ",
"nz-divider { display: block; } "],
                    encapsulation: ViewEncapsulation.None
                },] },
    ];
    /** @nocollapse */
    NzDividerComponent.ctorParameters = function () { return [
        { type: ElementRef, },
        { type: Renderer2, },
    ]; };
    NzDividerComponent.propDecorators = {
        "nzTitle": [{ type: Input },],
        "nzType": [{ type: Input },],
        "nzDashed": [{ type: Input },],
    };
    return NzDividerComponent;
}());
export { NzDividerComponent };
