// ngx-ueditor/index.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { UEditorComponent } from './src/ueditor.component';
import { UEditorConfig } from './src/ueditor.config';
import { ScriptService } from './src/script.service';
export { UEditorComponent } from './src/ueditor.component';
export { UEditorConfig } from './src/ueditor.config';
var UEditorModule = (function () {
    function UEditorModule() {
    }
    UEditorModule.forRoot = function (config) {
        return {
            ngModule: UEditorModule,
            providers: [
                { provide: UEditorConfig, useValue: config }
            ]
        };
    };
    UEditorModule.decorators = [
        { type: NgModule, args: [{
                    imports: [CommonModule],
                    providers: [ScriptService],
                    declarations: [UEditorComponent],
                    exports: [UEditorComponent]
                },] },
    ];
    /** @nocollapse */
    UEditorModule.ctorParameters = function () { return []; };
    return UEditorModule;
}());
export { UEditorModule };
