// @delon/abc/footer-toolbar/footer-toolbar.component.js
import { Component, Input, Inject, TemplateRef, ContentChild } from '@angular/core';
import { ElementRef, Renderer2 } from '@angular/core';
import { DOCUMENT } from '@angular/platform-browser';
import { coerceBooleanProperty } from '@angular/cdk/coercion';
var CLS = 'footer-toolbar';
var FooterToolbarComponent = /** @class */ (function () {
    function FooterToolbarComponent(el, renderer, doc) {
        this.el = el;
        this.renderer = renderer;
        this.doc = doc;
        this._errorCollect = false;
    }
    Object.defineProperty(FooterToolbarComponent.prototype, "errorCollect", {
        get: function () { return this._errorCollect; },
        set: function (value) {
            this._errorCollect = coerceBooleanProperty(value);
        },
        enumerable: true,
        configurable: true
    });
    FooterToolbarComponent.prototype.ngOnInit = function () {
        this.el.nativeElement.classList.add('footer-toolbar');
        this.renderer.addClass(this.el.nativeElement, CLS);
        this.doc.querySelector('body').classList.add("has-" + CLS);
    };
    FooterToolbarComponent.prototype.ngOnDestroy = function () {
        this.doc.querySelector('body').classList.remove("has-" + CLS);
    };
    FooterToolbarComponent.decorators = [
        { type: Component, args: [{
                    selector: 'footer-toolbar',
                    template: "\n    <div class=\"left\"><ng-container *ngIf=\"extra\" [ngTemplateOutlet]=\"extra\"></ng-container></div>\n    <div class=\"right\">\n        <error-collect *ngIf=\"errorCollect\"></error-collect>\n        <ng-content></ng-content>\n    </div>\n    "
                },] },
    ];
    /** @nocollapse */
    FooterToolbarComponent.ctorParameters = function () { return [
        { type: ElementRef, },
        { type: Renderer2, },
        { type: undefined, decorators: [{ type: Inject, args: [DOCUMENT,] },] },
    ]; };
    FooterToolbarComponent.propDecorators = {
        "errorCollect": [{ type: Input },],
        "extra": [{ type: ContentChild, args: ['extra',] },],
    };
    return FooterToolbarComponent;
}());
export { FooterToolbarComponent };
