// @delon/abc/desc-list/desc-list.component.js
import { Component, Input, ElementRef, Renderer2, TemplateRef, ContentChildren, QueryList } from '@angular/core';
import { coerceNumberProperty } from '@angular/cdk/coercion';
import { DescListItemComponent } from './desc-list-item.component';
var DescListComponent = /** @class */ (function () {
    // endregion
    function DescListComponent(el, renderer) {
        this.el = el;
        this.renderer = renderer;
        // region fields
        this._title = '';
        this._gutter = 32;
        this.layout = 'horizontal';
        this._xs = 24;
        this._sm = 12;
        this._md = 8;
        this._col = 3;
        this._classMap = [];
    }
    Object.defineProperty(DescListComponent.prototype, "title", {
        set: function (value) {
            if (value instanceof TemplateRef)
                this._titleTpl = value;
            else
                this._title = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DescListComponent.prototype, "gutter", {
        get: /** 列表项间距，单位为 `px` */
        function () { return this._gutter; },
        set: function (value) {
            this._gutter = coerceNumberProperty(value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DescListComponent.prototype, "col", {
        get: /** 指定信息最多分几列展示，最终一行几列由 col 配置结合响应式规则决定 */
        function () { return this._col; },
        set: function (value) {
            this._col = coerceNumberProperty(value);
        },
        enumerable: true,
        configurable: true
    });
    DescListComponent.prototype.setClass = function () {
        var _this = this;
        this._classMap.forEach(function (cls) { return _this.renderer.removeClass(_this.el.nativeElement, cls); });
        this._classMap = ['desc-list', this.layout];
        if (this.size)
            this._classMap.push('desc-list-' + this.size);
        this._classMap.forEach(function (cls) { return _this.renderer.addClass(_this.el.nativeElement, cls); });
    };
    DescListComponent.prototype.setResponsive = function () {
        var responsive = ({
            1: { xs: 24 },
            2: { xs: 24, sm: 12 },
            3: { xs: 24, sm: 12, md: 8 },
            4: { xs: 24, sm: 12, md: 6 },
        })[this.col > 4 ? 4 : this.col];
        this._xs = responsive.xs;
        this._sm = responsive.sm;
        this._md = responsive.md;
    };
    DescListComponent.prototype.ngOnInit = function () {
        this.setClass();
    };
    DescListComponent.prototype.ngOnChanges = function (changes) {
        if (changes.size && !changes.size.firstChange)
            this.setClass();
        if (changes.col)
            this.setResponsive();
    };
    DescListComponent.decorators = [
        { type: Component, args: [{
                    selector: 'desc-list',
                    template: "\n    <div *ngIf=\"_title || _titleTpl\" class=\"title\">\n        <ng-container *ngIf=\"_title; else _titleTpl\">{{_title}}</ng-container>\n    </div>\n    <div nz-row [nzGutter]=\"gutter\">\n        <div nz-col [nzXs]=\"_xs\" [nzSm]=\"_sm\" [nzMd]=\"_md\" *ngFor=\"let i of _items\">\n            <ng-template [ngTemplateOutlet]=\"i.tpl\"></ng-template>\n        </div>\n    </div>\n    ",
                    styles: ["/* stylelint-disable at-rule-empty-line-before,at-rule-name-space-after,at-rule-no-unknown */ /* stylelint-disable declaration-bang-space-before */ /* stylelint-disable declaration-bang-space-before */ :host { display: block; } :host ::ng-deep .title { font-size: 14px; color: rgba(0, 0, 0, 0.85); font-weight: 500; margin-bottom: 16px; } :host ::ng-deep .term { line-height: 22px; padding-bottom: 16px; margin-right: 8px; color: rgba(0, 0, 0, 0.85); white-space: nowrap; display: table-cell; } :host ::ng-deep .term:after { content: \"\:\"; margin: 0 8px 0 2px; position: relative; top: -0.5px; } :host ::ng-deep .detail { line-height: 22px; width: 100%; padding-bottom: 16px; color: rgba(0, 0, 0, 0.65); display: table-cell; } :host ::ng-deep .ant-row { margin-bottom: -16px; overflow: hidden; } :host.vertical ::ng-deep .term { padding-bottom: 8px; display: block; } :host.vertical ::ng-deep .detail { display: block; } :host.small ::ng-deep .title { margin-bottom: 12px; color: rgba(0, 0, 0, 0.65); } :host.small ::ng-deep .term, :host.small ::ng-deep .detail { padding-bottom: 8px; } :host.small ::ng-deep .ant-row { margin-bottom: -8px; } :host.large ::ng-deep .title { font-size: 16px; } "]
                },] },
    ];
    /** @nocollapse */
    DescListComponent.ctorParameters = function () { return [
        { type: ElementRef, },
        { type: Renderer2, },
    ]; };
    DescListComponent.propDecorators = {
        "title": [{ type: Input },],
        "size": [{ type: Input },],
        "gutter": [{ type: Input },],
        "layout": [{ type: Input },],
        "col": [{ type: Input },],
        "_items": [{ type: ContentChildren, args: [DescListItemComponent,] },],
    };
    return DescListComponent;
}());
export { DescListComponent };
