// @delon/theme/services/menu/menu.service.js
import { Injectable, Inject, Optional } from '@angular/core';
import { BehaviorSubject } from 'rxjs/BehaviorSubject';
import { share } from 'rxjs/operators';
import { ALAIN_I18N_TOKEN } from '../i18n/i18n';
import { ACLService } from '@delon/acl';
var MenuService = /** @class */ (function () {
    function MenuService(i18nService, aclService) {
        this.i18nService = i18nService;
        this.aclService = aclService;
        this._change$ = new BehaviorSubject([]);
        this.data = [];
    }
    Object.defineProperty(MenuService.prototype, "change", {
        get: function () {
            return this._change$.pipe(share());
        },
        enumerable: true,
        configurable: true
    });
    MenuService.prototype.visit = function (callback) {
        var inFn = function (list, parentMenu, depth) {
            for (var _i = 0, list_1 = list; _i < list_1.length; _i++) {
                var item = list_1[_i];
                callback(item, parentMenu, depth);
                if (item.children && item.children.length > 0) {
                    inFn(item.children, item, depth + 1);
                }
                else {
                    item.children = [];
                }
            }
        };
        inFn(this.data, null, 0);
    };
    MenuService.prototype.add = function (items) {
        (_a = this.data).push.apply(_a, items);
        this.resume();
        var _a;
    };
    /**
     * 重置菜单，可能I18N、用户权限变动时需要调用刷新
     */
    /**
         * 重置菜单，可能I18N、用户权限变动时需要调用刷新
         */
    MenuService.prototype.resume = /**
         * 重置菜单，可能I18N、用户权限变动时需要调用刷新
         */
    function (callback) {
        var _this = this;
        var i = 1;
        this.removeShortcut();
        var shortcuts = [];
        this.visit(function (item, parent, depth) {
            item.__id = i++;
            item.__parent = parent;
            item._depth = depth;
            if (!item.link)
                item.link = '';
            if (!item.externalLink)
                item.externalLink = '';
            // badge
            if (item.badge) {
                if (item.badge_dot !== true) {
                    item.badge_dot = false;
                }
                if (!item.badge_status) {
                    item.badge_status = 'error';
                }
            }
            item._type = item.externalLink ? 2 : 1;
            if (item.children && item.children.length > 0) {
                item._type = 3;
            }
            // shortcut
            if (item.shortcut === true && (item.link || item.externalLink))
                shortcuts.push(item);
            var i18n = item.i18n || item.translate;
            item.text = _this.i18nService && i18n ? _this.i18nService.fanyi(i18n) : item.text;
            // hidden
            item._hidden = typeof item.hide === 'undefined' ? false : item.hide;
            // acl
            if (item.acl && _this.aclService) {
                item._hidden = !_this.aclService.can(item.acl);
            }
            if (callback)
                callback(item, parent, depth);
        });
        this.loadShortcut(shortcuts);
        this._change$.next(this.data);
    };
    /**
     * 加载快捷菜单，加载位置规则如下：
     * 1、统一在下标0的节点下（即【主导航】节点下方）
     *      1、若 children 存在 【shortcut_root: true】则最优先【推荐】这种方式
     *      2、否则查找带有【dashboard】字样链接，若存在则在此菜单的下方创建快捷入口
     *      3、否则放在0节点位置
     */
    /**
         * 加载快捷菜单，加载位置规则如下：
         * 1、统一在下标0的节点下（即【主导航】节点下方）
         *      1、若 children 存在 【shortcut_root: true】则最优先【推荐】这种方式
         *      2、否则查找带有【dashboard】字样链接，若存在则在此菜单的下方创建快捷入口
         *      3、否则放在0节点位置
         */
    MenuService.prototype.loadShortcut = /**
         * 加载快捷菜单，加载位置规则如下：
         * 1、统一在下标0的节点下（即【主导航】节点下方）
         *      1、若 children 存在 【shortcut_root: true】则最优先【推荐】这种方式
         *      2、否则查找带有【dashboard】字样链接，若存在则在此菜单的下方创建快捷入口
         *      3、否则放在0节点位置
         */
    function (shortcuts) {
        if (shortcuts.length === 0 || this.data.length === 0)
            return;
        var ls = this.data[0].children || [];
        var pos = ls.findIndex(function (w) { return w.shortcut_root === true; });
        if (pos === -1) {
            pos = ls.findIndex(function (w) { return w.link.includes('dashboard') || w.externalLink.includes('dashboard'); });
            pos = (pos !== -1 ? pos : -1) + 1;
            this.data[0].children.splice(pos, 0, {
                text: '快捷菜单',
                translate: 'shortcut',
                icon: 'icon-rocket',
                children: []
            });
        }
        var _data = this.data[0].children[pos];
        _data = Object.assign(_data, {
            shortcut_root: true,
            _type: 3,
            __id: -1,
            _depth: 1
        });
        _data.children = shortcuts.map(function (i) {
            i._depth = 2;
            return i;
        });
    };
    MenuService.prototype.removeShortcut = function () {
        var ls = this.data && this.data.length && this.data[0].children || [];
        var pos = ls.findIndex(function (w) { return w.shortcut_root === true; });
        if (pos !== -1)
            ls.splice(pos, 1);
    };
    Object.defineProperty(MenuService.prototype, "menus", {
        get: function () {
            return this.data;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * 清空菜单
     */
    /**
         * 清空菜单
         */
    MenuService.prototype.clear = /**
         * 清空菜单
         */
    function () {
        this.data = [];
        this._change$.next(this.data);
    };
    /**
     * 根据URL设置菜单 `_open` 属性
     * @param url URL地址
     */
    /**
         * 根据URL设置菜单 `_open` 属性
         * @param url URL地址
         */
    MenuService.prototype.openedByUrl = /**
         * 根据URL设置菜单 `_open` 属性
         * @param url URL地址
         */
    function (url) {
        if (!url) {
            return;
        }
        var findItem = null;
        this.visit(function (item) {
            item._open = false;
            if (!item.link) {
                return;
            }
            if (!findItem && url.startsWith(item.link)) {
                findItem = item;
            }
        });
        if (!findItem) {
            console.warn("not found page name: " + url);
            return;
        }
        do {
            findItem._open = true;
            findItem = findItem.__parent;
        } while (findItem);
    };
    /**
     * 根据url获取菜单列表
     * @param url
     */
    /**
         * 根据url获取菜单列表
         * @param url
         */
    MenuService.prototype.getPathByUrl = /**
         * 根据url获取菜单列表
         * @param url
         */
    function (url) {
        var item = null;
        this.visit(function (i, parent, depth) {
            if (i.link === url) {
                item = i;
            }
        });
        var ret = [];
        if (!item)
            return ret;
        do {
            ret.splice(0, 0, item);
            item = item.__parent;
        } while (item);
        return ret;
    };
    MenuService.prototype.ngOnDestroy = function () {
        if (this._change$)
            this._change$.unsubscribe();
    };
    MenuService.decorators = [
        { type: Injectable },
    ];
    /** @nocollapse */
    MenuService.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: Optional }, { type: Inject, args: [ALAIN_I18N_TOKEN,] },] },
        { type: ACLService, decorators: [{ type: Optional },] },
    ]; };
    return MenuService;
}());
export { MenuService };
