// @delon/theme/services/colors/colors.service.js
import { Injectable } from '@angular/core';
var ColorsService = /** @class */ (function () {
    function ColorsService() {
        this.APP_COLORS = {
            primary: '#108ee9',
            success: '#00a854',
            error: '#f04134',
            warning: '#ffbf00',
            red: '#f04134',
            green: '#00a854',
            blue: '#108ee9',
            pink: '#f5317f',
            orange: '#f56a00',
            purple: '#7265e6',
            yellow: '#ffbf00',
            cyan: '#00a2ae',
            teal: '#20C997',
            grey: '#bfbfbf'
        };
    }
    ColorsService.prototype.byName = function (name) {
        return (this.APP_COLORS[name] || '#fff');
    };
    Object.defineProperty(ColorsService.prototype, "colors", {
        get: function () {
            return this.APP_COLORS;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ColorsService.prototype, "names", {
        get: function () {
            return ['red', 'green', 'blue', 'pink', 'orange', 'purple', 'yellow', 'cyan', 'teal', 'grey'];
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ColorsService.prototype, "brands", {
        get: function () {
            return ['primary', 'success', 'error', 'warning'];
        },
        enumerable: true,
        configurable: true
    });
    ColorsService.decorators = [
        { type: Injectable },
    ];
    /** @nocollapse */
    ColorsService.ctorParameters = function () { return []; };
    return ColorsService;
}());
export { ColorsService };
