// ngx-tinymce/bundles/ngx-tinymce.umd.js
(function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/common'), require('@angular/forms'), require('@angular/router'), require('rxjs/operators'), require('rxjs/Subject')) :
	typeof define === 'function' && define.amd ? define(['exports', '@angular/core', '@angular/common', '@angular/forms', '@angular/router', 'rxjs/operators', 'rxjs/Subject'], factory) :
	(factory((global.ngx = global.ngx || {}, global.ngx.tinymce = {}),global.ng.core,global.ng.common,global.ng.forms,global.ng.router,global.Rx.Observable.prototype,global.Rx));
}(this, (function (exports,core,common,forms,router,operators,Subject) { 'use strict';

var TinymceOptions = (function () {
    function TinymceOptions() {
        /** 指定tinymce目录路径，默认：`./assets/tinymce/` */
        this.baseURL = './assets/tinymce/';
        /** 指定tinymce文件名，默认：`tinymce.min.js` */
        this.fileName = 'tinymce.min.js';
    }
    return TinymceOptions;
}());

var ScriptService = (function () {
    function ScriptService(doc) {
        this.doc = doc;
        this.loaded = false;
        this.list = {};
        this.emitter = new Subject.Subject();
    }
    ScriptService.prototype.getChangeEmitter = function () {
        return this.emitter;
    };
    ScriptService.prototype.load = function (path) {
        var _this = this;
        if (this.loaded) {
            return this;
        }
        this.loaded = true;
        var promises = [];
        [path].forEach(function (script) { return promises.push(_this.loadScript(script)); });
        Promise.all(promises).then(function (res) {
            _this.emitter.next(true);
        });
        return this;
    };
    ScriptService.prototype.loadScript = function (path) {
        var _this = this;
        return new Promise(function (resolve, reject) {
            if (_this.list[path] === true) {
                resolve({
                    path: path,
                    loaded: true,
                    status: 'Loaded'
                });
                return;
            }
            _this.list[path] = true;
            var node = _this.doc.createElement('script');
            node.type = 'text/javascript';
            node.src = path;
            node.charset = 'utf-8';
            if (node.readyState) {
                // IE
                node.onreadystatechange = function () {
                    if (node.readyState === 'loaded' || node.readyState === 'complete') {
                        node.onreadystatechange = null;
                        resolve({
                            path: path,
                            loaded: true,
                            status: 'Loaded'
                        });
                    }
                };
            }
            else {
                node.onload = function () {
                    resolve({
                        path: path,
                        loaded: true,
                        status: 'Loaded'
                    });
                };
            }
            node.onerror = function (error) {
                return resolve({
                    path: path,
                    loaded: false,
                    status: 'Loaded'
                });
            };
            _this.doc.getElementsByTagName('head')[0].appendChild(node);
        });
    };
    ScriptService.decorators = [
        { type: core.Injectable },
    ];
    /** @nocollapse */
    ScriptService.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: core.Inject, args: [common.DOCUMENT,] },] },
    ]; };
    return ScriptService;
}());

var TinymceComponent = (function () {
    function TinymceComponent(defConfig, ss, router$$1, cd) {
        this.defConfig = defConfig;
        this.ss = ss;
        this.router = router$$1;
        this.cd = cd;
        this.inited = false;
        this.load = true;
        this.id = "_tinymce-" + Math.random().toString(36).substring(2);
        this.onChange = Function.prototype;
        this.onTouched = Function.prototype;
    }
    Object.defineProperty(TinymceComponent.prototype, "loading", {
        set: function (value) {
            if (value instanceof core.TemplateRef)
                this._loadingTpl = value;
            else
                this._loading = value;
        },
        enumerable: true,
        configurable: true
    });
    TinymceComponent.prototype.init = function () {
        var _this = this;
        if (!window.tinymce)
            throw new Error('tinymce js文件加载失败');
        if (this.instance)
            return;
        if (this.defConfig.baseURL)
            tinymce.baseURL = this.defConfig.baseURL;
        var userOptions = Object.assign({}, this.defConfig.config, this.config);
        var options = Object.assign({
            selector: "#" + this.id
        }, this.defConfig.config, this.config, {
            setup: function (editor) {
                _this.instance = editor;
                editor.on('change keyup', function () {
                    _this.value = editor.getContent();
                    _this.onChange(_this.value);
                    _this.onTouched(_this.value);
                });
                if (typeof userOptions.setup === 'function') {
                    userOptions.setup(editor);
                }
            },
            init_instance_callback: function (editor) {
                if (editor && _this.value)
                    editor.setContent(_this.value);
                if (typeof userOptions.init_instance_callback === 'function') {
                    userOptions.init_instance_callback(editor);
                }
                _this.load = false;
                _this.cd.markForCheck();
            }
        });
        if (userOptions.auto_focus)
            options.auto_focus = this.id;
        tinymce.init(options);
    };
    TinymceComponent.prototype.destroy = function () {
        if (!this.instance) {
            return;
        }
        this.instance.off();
        this.instance.remove('#' + this.id);
        this.instance = null;
    };
    TinymceComponent.prototype.ngOnInit = function () {
        var _this = this;
        this.inited = true;
        this.route$ = this.router.events.pipe(operators.filter(function (e) { return e instanceof router.ActivationEnd; }), operators.debounceTime(100), operators.filter(function (e) { return !!document.querySelector('#' + _this.id); })).subscribe(function (res) {
            _this.destroy();
            _this.init();
        });
    };
    TinymceComponent.prototype.ngAfterViewInit = function () {
        var _this = this;
        // 已经存在对象无须进入懒加载模式
        if (window.tinymce) {
            this.init();
            return;
        }
        var baseURL = this.defConfig && this.defConfig.baseURL;
        var fileName = this.defConfig && this.defConfig.fileName;
        this.ss.load((baseURL || './assets/tinymce/') + (fileName || 'tinymce.min.js')).getChangeEmitter().subscribe(function (res) {
            _this.init();
        });
    };
    TinymceComponent.prototype.ngOnChanges = function (changes) {
        if (this.inited && changes.config) {
            this.destroy();
            this.init();
        }
    };
    TinymceComponent.prototype.ngOnDestroy = function () {
        this.destroy();
        if (this.route$)
            this.route$.unsubscribe();
    };
    TinymceComponent.prototype.writeValue = function (value) {
        this.value = value;
        if (this.instance) {
            this.instance.setContent(this.value);
        }
    };
    TinymceComponent.prototype.registerOnChange = function (fn) { this.onChange = fn; };
    TinymceComponent.prototype.registerOnTouched = function (fn) { this.onTouched = fn; };
    TinymceComponent.prototype.setDisabledState = function (isDisabled) {
        if (!this.instance)
            return;
        if (isDisabled) {
            this.instance.disabled();
        }
        else {
            this.instance.setEnabled();
        }
    };
    TinymceComponent.decorators = [
        { type: core.Component, args: [{
                    // tslint:disable-next-line:component-selector
                    selector: 'tinymce',
                    template: "<textarea id=\"{{id}}\" class=\"tinymce-selector\"></textarea>\n    <div class=\"loading\" *ngIf=\"load\">\n        <ng-container *ngIf=\"_loading; else _loadingTpl\">{{_loading}}</ng-container>\n    </div>",
                    styles: [":host .tinymce-selector{display:none}"],
                    providers: [{
                            provide: forms.NG_VALUE_ACCESSOR,
                            useExisting: core.forwardRef(function () { return TinymceComponent; }),
                            multi: true
                        }],
                    changeDetection: core.ChangeDetectionStrategy.OnPush
                },] },
    ];
    /** @nocollapse */
    TinymceComponent.ctorParameters = function () { return [
        { type: TinymceOptions, },
        { type: ScriptService, },
        { type: router.Router, decorators: [{ type: core.Optional },] },
        { type: core.ChangeDetectorRef, },
    ]; };
    TinymceComponent.propDecorators = {
        "config": [{ type: core.Input },],
        "loading": [{ type: core.Input },],
    };
    return TinymceComponent;
}());

// endregion
var NgxTinymceModule = (function () {
    function NgxTinymceModule() {
    }
    NgxTinymceModule.forRoot = function (options) {
        return {
            ngModule: NgxTinymceModule,
            providers: [
                ScriptService,
                { provide: TinymceOptions, useValue: options }
            ]
        };
    };
    NgxTinymceModule.decorators = [
        { type: core.NgModule, args: [{
                    imports: [common.CommonModule],
                    declarations: [TinymceComponent],
                    exports: [TinymceComponent]
                },] },
    ];
    /** @nocollapse */
    NgxTinymceModule.ctorParameters = function () { return []; };
    return NgxTinymceModule;
}());

exports.NgxTinymceModule = NgxTinymceModule;
exports.TinymceComponent = TinymceComponent;
exports.TinymceOptions = TinymceOptions;

Object.defineProperty(exports, '__esModule', { value: true });

})));
