// @delon/abc/zip/zip.service.js
import { Injectable, Inject } from '@angular/core';
import { saveAs } from 'file-saver';
import { _HttpClient } from '@delon/theme';
import { LazyService } from '../utils/lazy.service';
import { DA_ZIP_CONFIG } from './interface';
var ZipService = /** @class */ (function () {
    function ZipService(config, _http, lazy) {
        this.config = config;
        this._http = _http;
        this.lazy = lazy;
    }
    ZipService.prototype.init = function () {
        var config = Object.assign({
            url: "//cdn.bootcss.com/jszip/3.1.5/jszip.min.js",
            utils: []
        }, this.config);
        return this.lazy.load([config.url].concat(config.utils));
    };
    ZipService.prototype.check = function (zip) {
        if (!zip)
            throw new Error('get instance via `ZipService.create()`');
    };
    /** 解压 */
    /** 解压 */
    ZipService.prototype.read = /** 解压 */
    function (fileOrUrl, options) {
        var _this = this;
        return new Promise(function (resolve) {
            _this.init().then(function () {
                // from url
                if (typeof fileOrUrl === 'string') {
                    _this._http.request('GET', fileOrUrl, { responseType: 'arraybuffer' }).subscribe(function (res) {
                        JSZip.loadAsync(res, options).then(function (ret) { return resolve(ret); });
                    });
                    return;
                }
                // from file
                var reader = new FileReader();
                reader.onload = function (e) {
                    JSZip.loadAsync(e.target.result, options).then(function (ret) { return resolve(ret); });
                };
                reader.readAsBinaryString(fileOrUrl);
            });
        });
    };
    /** 创建 Zip 实例，用于创建压缩文件 */
    /** 创建 Zip 实例，用于创建压缩文件 */
    ZipService.prototype.create = /** 创建 Zip 实例，用于创建压缩文件 */
    function () {
        var _this = this;
        return new Promise(function (resolve) {
            _this.init().then(function () {
                var zipFile = new JSZip();
                resolve(zipFile);
            });
        });
    };
    /**
     * 下载URL资源并写入 zip
     * @param zip Zip 实例
     * @param path Zip 路径，例如： `text.txt`、`txt/hi.txt`
     * @param url URL 地址
     */
    /**
         * 下载URL资源并写入 zip
         * @param zip Zip 实例
         * @param path Zip 路径，例如： `text.txt`、`txt/hi.txt`
         * @param url URL 地址
         */
    ZipService.prototype.pushUrl = /**
         * 下载URL资源并写入 zip
         * @param zip Zip 实例
         * @param path Zip 路径，例如： `text.txt`、`txt/hi.txt`
         * @param url URL 地址
         */
    function (zip, path, url) {
        var _this = this;
        this.check(zip);
        return new Promise(function (resolve, reject) {
            _this._http.request('GET', url, { responseType: 'arraybuffer' }).subscribe(function (res) {
                zip.file(path, res);
                resolve();
            }, function () {
                reject();
            });
        });
    };
    /** 保存Zip */
    /** 保存Zip */
    ZipService.prototype.save = /** 保存Zip */
    function (zip, options) {
        this.check(zip);
        var opt = Object.assign({}, options);
        return new Promise(function (resolve) {
            zip.generateAsync(Object.assign({ type: 'blob' }, opt.options), opt.update).then(function (data) {
                if (opt.callback)
                    opt.callback();
                saveAs(data, opt.filename || 'download.zip');
            });
        });
    };
    ZipService.decorators = [
        { type: Injectable },
    ];
    /** @nocollapse */
    ZipService.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: Inject, args: [DA_ZIP_CONFIG,] },] },
        { type: _HttpClient, },
        { type: LazyService, },
    ]; };
    return ZipService;
}());
export { ZipService };
