// @delon/abc/zip/zip.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { AdUtilsModule } from '../utils/utils.module';
import { ZipService } from './zip.service';
import { DA_ZIP_CONFIG } from './interface';
var COMPONENTS = [];
// region: zorro modules
// import { NzBreadCrumbModule } from 'ng-zorro-antd';
// import { NzIconModule } from 'ng-zorro-antd-extra';
var ZORROMODULES = [];
// endregion
var AdZipModule = /** @class */ (function () {
    function AdZipModule() {
    }
    AdZipModule.forRoot = function (config) {
        return {
            ngModule: AdZipModule,
            providers: [
                ZipService,
                { provide: DA_ZIP_CONFIG, useValue: config }
            ]
        };
    };
    AdZipModule.decorators = [
        { type: NgModule, args: [{
                    imports: [CommonModule].concat(ZORROMODULES, [AdUtilsModule]),
                    declarations: COMPONENTS.slice(),
                    exports: COMPONENTS.slice()
                },] },
    ];
    /** @nocollapse */
    AdZipModule.ctorParameters = function () { return []; };
    return AdZipModule;
}());
export { AdZipModule };
