// ng-zorro-antd-extra/nz-list/nz-list-item-meta.component.js
import { Component, Input, TemplateRef, HostBinding } from '@angular/core';
var NzListItemMetaComponent = (function () {
    function NzListItemMetaComponent() {
        this._nzListItemMeta = true;
        this.avatar = false;
        this._avatar = '';
        this.title = false;
        this._title = '';
        this.desc = false;
        this._desc = '';
    }
    Object.defineProperty(NzListItemMetaComponent.prototype, "nzAvatar", {
        set: function (value) {
            if (value instanceof TemplateRef)
                this._avatarTpl = value;
            else
                this._avatar = value;
            this.avatar = !!value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NzListItemMetaComponent.prototype, "nzTitle", {
        set: function (value) {
            if (value instanceof TemplateRef)
                this._titleTpl = value;
            else
                this._title = value;
            this.title = !!value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NzListItemMetaComponent.prototype, "nzDescription", {
        set: function (value) {
            if (value instanceof TemplateRef)
                this._descTpl = value;
            else
                this._desc = value;
            this.desc = !!value;
        },
        enumerable: true,
        configurable: true
    });
    NzListItemMetaComponent.decorators = [
        { type: Component, args: [{
                    selector: 'nz-list-item-meta',
                    template: "\n    <div *ngIf=\"avatar\" class=\"ant-list-item-meta-avatar\">\n        <ng-container *ngIf=\"_avatar; else _avatarTpl\"><nz-avatar [nzSrc]=\"_avatar\"></nz-avatar></ng-container>\n    </div>\n    <div *ngIf=\"title || desc\" class=\"ant-list-item-meta-content\">\n        <h4 *ngIf=\"title\" class=\"ant-list-item-meta-title\">\n            <ng-container *ngIf=\"_title; else _titleTpl\">{{ _title }}</ng-container>\n        </h4>\n        <div *ngIf=\"desc\" class=\"ant-list-item-meta-description\">\n            <ng-container *ngIf=\"_desc; else _descTpl\">{{ _desc }}</ng-container>\n        </div>\n    </div>\n    "
                },] },
    ];
    /** @nocollapse */
    NzListItemMetaComponent.ctorParameters = function () { return []; };
    NzListItemMetaComponent.propDecorators = {
        "_nzListItemMeta": [{ type: HostBinding, args: ['class.ant-list-item-meta',] },],
        "nzAvatar": [{ type: Input },],
        "nzTitle": [{ type: Input },],
        "nzDescription": [{ type: Input },],
    };
    return NzListItemMetaComponent;
}());
export { NzListItemMetaComponent };
