// @delon/abc/tag-select/tag-select.component.js
import { Component, HostBinding, Input, Output, EventEmitter } from '@angular/core';
import { coerceBooleanProperty } from '@angular/cdk/coercion';
var TagSelectComponent = /** @class */ (function () {
    function TagSelectComponent() {
        this._cls = true;
        this._expandable = true;
        this.expand = false;
        this.change = new EventEmitter();
    }
    Object.defineProperty(TagSelectComponent.prototype, "expandable", {
        get: /** 是否启用 `展开与收进` */
        function () { return this._expandable; },
        set: function (value) {
            this._expandable = coerceBooleanProperty(value);
        },
        enumerable: true,
        configurable: true
    });
    TagSelectComponent.prototype.trigger = function () {
        this.expand = !this.expand;
        this.change.emit(this.expand);
    };
    TagSelectComponent.decorators = [
        { type: Component, args: [{
                    selector: 'tag-select',
                    template: "\n    <ng-content></ng-content>\n    <a *ngIf=\"expandable\" class=\"trigger\" (click)=\"trigger()\">{{expand ? '\u6536\u8D77' : '\u5C55\u5F00'}}<i class=\"anticon anticon-{{expand ? 'up' : 'down'}} ml-sm\"></i></a>",
                    styles: ["/* stylelint-disable at-rule-empty-line-before,at-rule-name-space-after,at-rule-no-unknown */ /* stylelint-disable declaration-bang-space-before */ /* stylelint-disable declaration-bang-space-before */ :host { user-select: none; margin-left: -8px; position: relative; overflow: hidden; max-height: 32px; line-height: 32px; transition: all 0.3s; display: block; } :host.expanded { transition: all 0.3s; max-height: 200px; } :host.has-expand { padding-right: 50px; } :host ::ng-deep .ant-tag { padding: 0 8px; margin-right: 24px; } :host ::ng-deep .trigger { position: absolute; top: 0; right: 0; } :host ::ng-deep .trigger i { font-size: 12px; } "]
                },] },
    ];
    /** @nocollapse */
    TagSelectComponent.ctorParameters = function () { return []; };
    TagSelectComponent.propDecorators = {
        "_cls": [{ type: HostBinding, args: ['class.tag-select',] },],
        "expandable": [{ type: HostBinding, args: ['class.has-expand',] }, { type: Input },],
        "expand": [{ type: HostBinding, args: ['class.expanded',] },],
        "change": [{ type: Output },],
    };
    return TagSelectComponent;
}());
export { TagSelectComponent };
