// @delon/abc/exception/exception.component.js
import { Component, Input, ContentChild, TemplateRef } from '@angular/core';
var ExceptionComponent = /** @class */ (function () {
    function ExceptionComponent() {
        this._img = '';
        this._title = '';
        this._desc = '';
    }
    Object.defineProperty(ExceptionComponent.prototype, "type", {
        set: function (value) {
            var item = ({
                403: {
                    img: 'https://gw.alipayobjects.com/zos/rmsportal/wZcnGqRDyhPOEYFcZDnb.svg',
                    title: '403',
                    desc: '抱歉，你无权访问该页面',
                },
                404: {
                    img: 'https://gw.alipayobjects.com/zos/rmsportal/KpnpchXsobRgLElEozzI.svg',
                    title: '404',
                    desc: '抱歉，你访问的页面不存在',
                },
                500: {
                    img: 'https://gw.alipayobjects.com/zos/rmsportal/RVRUAYdCGeYNBWoKiIwB.svg',
                    title: '500',
                    desc: '抱歉，服务器出错了',
                },
            })[value];
            if (!item)
                return;
            this._img = item.img;
            this._title = item.title;
            this._desc = item.desc;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ExceptionComponent.prototype, "img", {
        set: function (value) { this._img = value; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ExceptionComponent.prototype, "title", {
        set: function (value) { this._title = value; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ExceptionComponent.prototype, "desc", {
        set: function (value) { this._desc = value; },
        enumerable: true,
        configurable: true
    });
    ExceptionComponent.decorators = [
        { type: Component, args: [{
                    selector: 'exception',
                    template: "\n    <div class=\"img-block\">\n        <div class=\"img\" [ngStyle]=\"{'background-image': 'url(' + _img + ')'}\"></div>\n    </div>\n    <div class=\"cont\">\n        <h1 [innerHTML]=\"_title\"></h1>\n        <div class=\"desc\" [innerHTML]=\"_desc\"></div>\n        <ng-template #defaultActions>\n            <button nz-button [routerLink]=\"['/']\" [nzType]=\"'primary'\" [nzSize]=\"'large'\">\u8FD4\u56DE\u9996\u9875</button>\n            <ng-content></ng-content>\n        </ng-template>\n        <div class=\"actions\" *ngIf=\"actions; else defaultActions\">\n            <ng-template [ngTemplateOutlet]=\"actions\"></ng-template>\n            <ng-content></ng-content>\n        </div>\n    </div>\n    ",
                    styles: ["/* stylelint-disable at-rule-empty-line-before,at-rule-name-space-after,at-rule-no-unknown */ /* stylelint-disable declaration-bang-space-before */ /* stylelint-disable declaration-bang-space-before */ :host { display: flex; align-items: center; height: 100%; } :host ::ng-deep .img-block { flex: 0 0 62.5%; width: 62.5%; padding-right: 152px; zoom: 1; } :host ::ng-deep .img-block:before, :host ::ng-deep .img-block:after { content: \"\ \"; display: table; } :host ::ng-deep .img-block:after { clear: both; visibility: hidden; font-size: 0; height: 0; } :host ::ng-deep .img { height: 360px; width: 100%; max-width: 430px; float: right; background-repeat: no-repeat; background-position: 50% 50%; background-size: 100% 100%; } :host ::ng-deep .cont { flex: auto; } :host ::ng-deep .cont h1 { color: #434e59; font-size: 72px; font-weight: 600; line-height: 72px; margin-bottom: 24px; } :host ::ng-deep .cont .desc { color: rgba(0, 0, 0, 0.43); font-size: 20px; line-height: 28px; margin-bottom: 16px; } :host ::ng-deep .cont .actions button:not(:last-child) { margin-right: 8px; } @media screen and (max-width: 1600px) { :host ::ng-deep .img-block { padding-right: 88px; } } @media screen and (max-width: 768px) { :host { display: block; text-align: center; } :host ::ng-deep .img-block { padding-right: 0; margin: 0 auto 24px; } } @media screen and (max-width: 480px) { :host ::ng-deep .img-block { margin-bottom: -24px; overflow: hidden; } } "]
                },] },
    ];
    /** @nocollapse */
    ExceptionComponent.ctorParameters = function () { return []; };
    ExceptionComponent.propDecorators = {
        "type": [{ type: Input },],
        "img": [{ type: Input },],
        "title": [{ type: Input },],
        "desc": [{ type: Input },],
        "actions": [{ type: ContentChild, args: ['actions',] },],
    };
    return ExceptionComponent;
}());
export { ExceptionComponent };
