// @delon/theme/services/settings/settings.service.js
import { Injectable } from '@angular/core';
var KEY = 'layout';
var SettingsService = /** @class */ (function () {
    function SettingsService() {
        this.app = {
            year: (new Date()).getFullYear()
        };
        this.user = {};
        this._layout = null;
    }
    SettingsService.prototype.get = function (key) {
        return JSON.parse(localStorage.getItem(key) || 'null') || null;
    };
    SettingsService.prototype.set = function (key, value) {
        localStorage.setItem(key, JSON.stringify(value));
    };
    Object.defineProperty(SettingsService.prototype, "layout", {
        get: function () {
            if (!this._layout) {
                this._layout = Object.assign({
                    fixed: true,
                    collapsed: false,
                    boxed: false,
                    theme: 'A',
                    lang: null
                }, this.get(KEY));
                this.set(KEY, this._layout);
            }
            return this._layout;
        },
        enumerable: true,
        configurable: true
    });
    SettingsService.prototype.setLayout = function (name, value) {
        if (typeof this.layout[name] !== 'undefined') {
            this.layout[name] = value;
            this.set(KEY, this._layout);
            return true;
        }
        return false;
    };
    SettingsService.prototype.setApp = function (val) {
        this.app = Object.assign(this.app, val);
    };
    SettingsService.prototype.setUser = function (val) {
        this.user = Object.assign(this.user, val);
    };
    SettingsService.decorators = [
        { type: Injectable },
    ];
    /** @nocollapse */
    SettingsService.ctorParameters = function () { return []; };
    return SettingsService;
}());
export { SettingsService };
