// @delon/theme/index.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { RouterModule } from '@angular/router';
// region: import
import { ALAIN_THEME_OPTIONS } from './theme.options';
import { MenuService } from './services/menu/menu.service';
import { ColorsService } from './services/colors/colors.service';
import { ScrollService } from './services/scroll/scroll.service';
import { SettingsService } from './services/settings/settings.service';
import { TitleService } from './services/title/title.service';
import { ThemesService } from './services/themes/themes.service';
import { ALAIN_I18N_TOKEN, AlainI18NServiceFake } from './services/i18n/i18n';
import { _HttpClient } from './services/http/http.client';
var SERVICES = [ColorsService, MenuService, ScrollService, SettingsService, ThemesService, TitleService, _HttpClient];
import { ModalHelper } from './services/modal/modal.helper';
var HELPERS = [ModalHelper];
// components
var COMPONENTS = [];
// pipes
import { MomentDatePipe } from './pipes/date/moment-date.pipe';
import { CNCurrencyPipe } from './pipes/currency/cn-currency.pipe';
import { KeysPipe } from './pipes/keys/keys.pipe';
import { YNPipe } from './pipes/yn/yn.pipe';
var PIPES = [MomentDatePipe, CNCurrencyPipe, KeysPipe, YNPipe];
// endregion
// region: export
export { ALAIN_THEME_OPTIONS } from './theme.options';
export { preloaderFinished } from './services/preloader/preloader';
export { MenuService } from './services/menu/menu.service';
export { ColorsService } from './services/colors/colors.service';
export { ScrollService } from './services/scroll/scroll.service';
export { SettingsService } from './services/settings/settings.service';
export { TitleService } from './services/title/title.service';
export { ThemesService } from './services/themes/themes.service';
export { ALAIN_I18N_TOKEN } from './services/i18n/i18n';
export { ModalHelper } from './services/modal/modal.helper';
export { _HttpClient } from './services/http/http.client';
export { MomentDatePipe } from './pipes/date/moment-date.pipe';
export { CNCurrencyPipe } from './pipes/currency/cn-currency.pipe';
export { KeysPipe } from './pipes/keys/keys.pipe';
export { YNPipe } from './pipes/yn/yn.pipe';
// endregion
// region: zorro modules
import { NzToolTipModule } from 'ng-zorro-antd';
var ZORROMODULES = [NzToolTipModule];
// endregion
var AlainThemeModule = /** @class */ (function () {
    function AlainThemeModule() {
    }
    AlainThemeModule.forRoot = function (options) {
        return {
            ngModule: AlainThemeModule,
            providers: [
                { provide: ALAIN_THEME_OPTIONS, useValue: options || {} },
                { provide: ALAIN_I18N_TOKEN, useClass: AlainI18NServiceFake }
            ].concat(SERVICES, HELPERS)
        };
    };
    AlainThemeModule.forChild = function () {
        return {
            ngModule: AlainThemeModule,
            providers: HELPERS.slice()
        };
    };
    AlainThemeModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule,
                        RouterModule
                    ].concat(ZORROMODULES),
                    declarations: COMPONENTS.concat(PIPES),
                    exports: COMPONENTS.concat(PIPES)
                },] },
    ];
    /** @nocollapse */
    AlainThemeModule.ctorParameters = function () { return []; };
    return AlainThemeModule;
}());
export { AlainThemeModule };
