// @delon/abc/xlsx/xlsx.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { AdUtilsModule } from '../utils/utils.module';
import { XlsxService } from './xlsx.service';
import { XlsxDirective } from './xlsx.directive';
import { DA_XLSX_CONFIG } from './interface';
var COMPONENTS = [XlsxDirective];
// region: zorro modules
// import { NzBreadCrumbModule } from 'ng-zorro-antd';
// import { NzIconModule } from 'ng-zorro-antd-extra';
var ZORROMODULES = [];
// endregion
var AdXlsxModule = /** @class */ (function () {
    function AdXlsxModule() {
    }
    AdXlsxModule.forRoot = function (config) {
        return {
            ngModule: AdXlsxModule,
            providers: [
                XlsxService,
                { provide: DA_XLSX_CONFIG, useValue: config }
            ]
        };
    };
    AdXlsxModule.decorators = [
        { type: NgModule, args: [{
                    imports: [CommonModule].concat(ZORROMODULES, [AdUtilsModule]),
                    declarations: COMPONENTS.slice(),
                    exports: COMPONENTS.slice()
                },] },
    ];
    /** @nocollapse */
    AdXlsxModule.ctorParameters = function () { return []; };
    return AdXlsxModule;
}());
export { AdXlsxModule };
