// ng-clipboard-antd/src/directive.js
import { Directive, Input, EventEmitter, Output, HostListener } from '@angular/core';
import { NzClipboardService } from './service';
var NzClipboardDirective = (function () {
    function NzClipboardDirective(clipboardSrv) {
        this.clipboardSrv = clipboardSrv;
        this.success = new EventEmitter();
        this.error = new EventEmitter();
    }
    NzClipboardDirective.prototype._click = function () {
        if (!this.clipboardSrv.isSupported) {
            this.handleResult(false, undefined);
        }
        else if (this.target) {
            var newTarget = null;
            if (!(this.target instanceof HTMLInputElement)) {
                newTarget = this.target._el.querySelector('input,textarea');
            }
            else {
                newTarget = this.target;
            }
            this.clipboardSrv.isTargetValid(newTarget);
            this.handleResult(this.clipboardSrv.copyFromInputElement(newTarget, this.successText, this.errorText), newTarget.value);
        }
        else if (this.content) {
            this.handleResult(this.clipboardSrv.copyFromContent(this.content, this.successText, this.errorText), this.content);
        }
    };
    NzClipboardDirective.prototype.handleResult = function (succeeded, copiedContent) {
        if (succeeded) {
            this.success.emit({ status: true, content: copiedContent });
        }
        else {
            this.error.emit({ status: false });
        }
    };
    NzClipboardDirective.prototype.ngOnDestroy = function () {
        this.clipboardSrv.destroy();
    };
    NzClipboardDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[nz-clipboard]'
                },] },
    ];
    /** @nocollapse */
    NzClipboardDirective.ctorParameters = function () { return [
        { type: NzClipboardService, },
    ]; };
    NzClipboardDirective.propDecorators = {
        "content": [{ type: Input, args: ['nz-clipboard',] },],
        "target": [{ type: Input, args: ['nzTarget',] },],
        "successText": [{ type: Input, args: ['nzSuccessText',] },],
        "errorText": [{ type: Input, args: ['nzErrorText',] },],
        "success": [{ type: Output, args: ['nzSuccess',] },],
        "error": [{ type: Output, args: ['nzError',] },],
        "_click": [{ type: HostListener, args: ['click', ['$event.target'],] },],
    };
    return NzClipboardDirective;
}());
export { NzClipboardDirective };
