// ng-clipboard-antd/src/service.js
/**
 * The core function is ported from clipboard.js & ngx-clipboard
 * clipboard.js: https://github.com/zenorocha/clipboard.js
 * ngx-clipboard: https://github.com/maxisam/ngx-clipboard
 */
import { Injectable, Inject } from '@angular/core';
import { DOCUMENT } from '@angular/common';
import { NzMessageService } from 'ng-zorro-antd';
import { NzClipboardConfig } from './config';
var NzClipboardService = (function () {
    function NzClipboardService(document, cog, messageSrv) {
        this.document = document;
        this.messageSrv = messageSrv;
        this.config = Object.assign(new NzClipboardConfig(), cog);
    }
    Object.defineProperty(NzClipboardService.prototype, "isSupported", {
        get: function () {
            return !!this.document.queryCommandSupported && !!this.document.queryCommandSupported('copy');
        },
        enumerable: true,
        configurable: true
    });
    NzClipboardService.prototype.isTargetValid = function (element) {
        if (element instanceof HTMLInputElement || element instanceof HTMLTextAreaElement) {
            if (element.hasAttribute('disabled')) {
                // tslint:disable-next-line:max-line-length
                throw new Error('Invalid "target" attribute. Please use "readonly" instead of "disabled" attribute');
            }
            return true;
        }
        throw new Error('Target should be input or textarea');
    };
    /**
     * copyFromInputElement
     */
    /**
         * copyFromInputElement
         */
    NzClipboardService.prototype.copyFromInputElement = /**
         * copyFromInputElement
         */
    function (targetElm, success, error) {
        try {
            this.selectTarget(targetElm);
            var re = this.copyText();
            this.clearSelection(targetElm);
            this.handleMessage(true, success);
            return re;
        }
        catch (error) {
            console.warn('copy error', error);
            this.handleMessage(false, error);
            return false;
        }
    };
    /**
     * Creates a fake textarea element, sets its value from `text` property,
     * and makes a selection on it.
     */
    /**
         * Creates a fake textarea element, sets its value from `text` property,
         * and makes a selection on it.
         */
    NzClipboardService.prototype.copyFromContent = /**
         * Creates a fake textarea element, sets its value from `text` property,
         * and makes a selection on it.
         */
    function (content, success, error) {
        if (!this.tempTextArea) {
            this.tempTextArea = this.createTempTextArea(this.document);
            this.document.body.appendChild(this.tempTextArea);
        }
        this.tempTextArea.value = content;
        return this.copyFromInputElement(this.tempTextArea, success, error);
    };
    // remove temporary textarea if any
    // remove temporary textarea if any
    NzClipboardService.prototype.destroy = 
    // remove temporary textarea if any
    function () {
        if (this.tempTextArea) {
            this.document.body.removeChild(this.tempTextArea);
            this.tempTextArea = undefined;
        }
    };
    // select the target html input element
    // select the target html input element
    NzClipboardService.prototype.selectTarget = 
    // select the target html input element
    function (inputElement) {
        if (inputElement.select)
            inputElement.select();
        if (inputElement.setSelectionRange)
            inputElement.setSelectionRange(0, inputElement.value.length);
        return inputElement.value.length;
    };
    NzClipboardService.prototype.copyText = function () {
        return this.document.execCommand('copy');
    };
    // Removes current selection and focus from `target` element.
    // Removes current selection and focus from `target` element.
    NzClipboardService.prototype.clearSelection = 
    // Removes current selection and focus from `target` element.
    function (inputElement) {
        // tslint:disable-next-line:no-unused-expression
        inputElement && inputElement.blur();
        window.getSelection().removeAllRanges();
    };
    // create a fake textarea for copy command
    // create a fake textarea for copy command
    NzClipboardService.prototype.createTempTextArea = 
    // create a fake textarea for copy command
    function (doc) {
        var isRTL = doc.documentElement.getAttribute('dir') === 'rtl';
        var ta;
        ta = doc.createElement('textarea');
        // Prevent zooming on iOS
        ta.style.fontSize = '12pt';
        // Reset box model
        ta.style.border = '0';
        ta.style.padding = '0';
        ta.style.margin = '0';
        // Move element out of screen horizontally
        ta.style.position = 'absolute';
        ta.style[isRTL ? 'right' : 'left'] = '-9999px';
        // Move element to the same position vertically
        var yPosition = window.pageYOffset || doc.documentElement.scrollTop;
        ta.style.top = yPosition + 'px';
        ta.setAttribute('readonly', '');
        return ta;
    };
    NzClipboardService.prototype.handleMessage = function (status, focus) {
        if (typeof focus === 'undefined') {
            focus = status ? this.config.success : this.config.error;
        }
        if (status) {
            if (focus)
                this.messageSrv.success(focus);
        }
        else {
            if (focus)
                this.messageSrv.error(focus);
        }
    };
    NzClipboardService.decorators = [
        { type: Injectable },
    ];
    /** @nocollapse */
    NzClipboardService.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: Inject, args: [DOCUMENT,] },] },
        { type: NzClipboardConfig, },
        { type: NzMessageService, },
    ]; };
    return NzClipboardService;
}());
export { NzClipboardService };
