// angular-baidu-maps/core/abm.component.js
import { Component, Input, ElementRef, EventEmitter, Output, NgZone, ViewEncapsulation } from '@angular/core';
import { LoaderService } from './loader.service';
import { AbmConfig } from './abm.config';
var AbmComponent = (function () {
    function AbmComponent(el, COG, loader, zone) {
        this.el = el;
        this.COG = COG;
        this.loader = loader;
        this.zone = zone;
        this.options = {};
        this.ready = new EventEmitter();
        this.map = null;
    }
    AbmComponent.prototype.ngOnInit = function () {
        this._initMap();
    };
    AbmComponent.prototype.ngOnChanges = function (changes) {
        if ('options' in changes)
            this._updateOptions();
    };
    AbmComponent.prototype._initMap = function () {
        var _this = this;
        if (this.map)
            return;
        this.loader.load().then(function () {
            _this.zone.runOutsideAngular(function () {
                try {
                    _this.map = new BMap.Map(_this.el.nativeElement, _this.options);
                }
                catch (ex) {
                    console.warn('地图初始化失败', ex);
                }
            });
            _this.ready.emit(_this.map);
        }).catch(function (error) {
            console.warn('js加载失败', error);
        });
    };
    AbmComponent.prototype._updateOptions = function () {
        this.options = Object.assign({}, this.COG.mapOptions, this.options);
        if (this.map) {
            this.map.setOptions(this.options);
        }
    };
    AbmComponent.prototype.destroy = function () {
        if (this.map) {
            this.map.clearOverlays();
            this.map.clearHotspots();
        }
    };
    AbmComponent.prototype.ngOnDestroy = function () {
        this.destroy();
    };
    AbmComponent.decorators = [
        { type: Component, args: [{
                    selector: 'abm-map',
                    template: "",
                    styles: ["\n        .angular-baidu-maps-container { display:block; width:100%; height:100%; }\n    "],
                    host: {
                        '[class.angular-baidu-maps-container]': 'true'
                    },
                    encapsulation: ViewEncapsulation.None
                },] },
    ];
    /** @nocollapse */
    AbmComponent.ctorParameters = function () { return [
        { type: ElementRef, },
        { type: AbmConfig, },
        { type: LoaderService, },
        { type: NgZone, },
    ]; };
    AbmComponent.propDecorators = {
        "options": [{ type: Input },],
        "ready": [{ type: Output },],
    };
    return AbmComponent;
}());
export { AbmComponent };
