// @delon/abc/reuse-tab/reuse-tab.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { RouterModule, RouteReuseStrategy } from '@angular/router';
import { ReuseTabComponent } from './reuse-tab.component';
import { ReuseTabService } from './reuse-tab.service';
import { ReuseTabStrategy } from './reuse-tab.strategy';
var COMPONENTS = [ReuseTabComponent];
// region: zorro modules
import { NzTabsModule, NzButtonModule, NzPopconfirmModule } from 'ng-zorro-antd';
var ZORROMODULES = [NzTabsModule, NzButtonModule, NzPopconfirmModule];
// endregion
var AdReuseTabModule = /** @class */ (function () {
    function AdReuseTabModule() {
    }
    AdReuseTabModule.forRoot = function () {
        return {
            ngModule: AdReuseTabModule,
            providers: [
                ReuseTabService,
                { provide: RouteReuseStrategy, useClass: ReuseTabStrategy, deps: [ReuseTabService] }
            ]
        };
    };
    AdReuseTabModule.decorators = [
        { type: NgModule, args: [{
                    imports: [CommonModule, RouterModule].concat(ZORROMODULES),
                    declarations: COMPONENTS.slice(),
                    exports: COMPONENTS.slice()
                },] },
    ];
    /** @nocollapse */
    AdReuseTabModule.ctorParameters = function () { return []; };
    return AdReuseTabModule;
}());
export { AdReuseTabModule };
