// angular-qq-maps/index.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { AqmComponent } from './core/aqm.component';
import { AqmPanoramaComponent } from './core/aqm-panorama.component';
import { AqmConfig } from './core/aqm.config';
import { LoaderService } from './core/loader.service';
export { AqmConfig } from './core/aqm.config';
export { AqmComponent } from './core/aqm.component';
export { AqmPanoramaComponent } from './core/aqm-panorama.component';
var AqmModule = (function () {
    function AqmModule() {
    }
    AqmModule.forRoot = function (config) {
        return {
            ngModule: AqmModule,
            providers: [
                { provide: AqmConfig, useValue: config }
            ]
        };
    };
    AqmModule.decorators = [
        { type: NgModule, args: [{
                    imports: [CommonModule],
                    providers: [LoaderService],
                    declarations: [AqmComponent, AqmPanoramaComponent],
                    exports: [AqmComponent, AqmPanoramaComponent]
                },] },
    ];
    /** @nocollapse */
    AqmModule.ctorParameters = function () { return []; };
    return AqmModule;
}());
export { AqmModule };
