// ng-zorro-antd-extra/nz-card/nz-card-meta.component.js
import { Component, Input, TemplateRef, HostBinding } from '@angular/core';
var NzCardMetaComponent = (function () {
    function NzCardMetaComponent() {
        this._nzListItemMeta = true;
        this.avatar = false;
        this._avatar = '';
        this.title = false;
        this._title = '';
        this.desc = false;
        this._desc = '';
    }
    Object.defineProperty(NzCardMetaComponent.prototype, "nzAvatar", {
        set: function (value) {
            if (value instanceof TemplateRef)
                this._avatarTpl = value;
            else
                this._avatar = value;
            this.avatar = !!value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NzCardMetaComponent.prototype, "nzTitle", {
        set: function (value) {
            if (value instanceof TemplateRef)
                this._titleTpl = value;
            else
                this._title = value;
            this.title = !!value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NzCardMetaComponent.prototype, "nzDescription", {
        set: function (value) {
            if (value instanceof TemplateRef)
                this._descTpl = value;
            else
                this._desc = value;
            this.desc = !!value;
        },
        enumerable: true,
        configurable: true
    });
    NzCardMetaComponent.decorators = [
        { type: Component, args: [{
                    selector: 'nz-card-meta',
                    template: "\n    <div class=\"ant-card-meta-content\">\n        <div *ngIf=\"avatar\" class=\"ant-card-meta-avatar\">\n            <ng-container *ngIf=\"_avatar; else _avatarTpl\"><nz-avatar [nzSrc]=\"_avatar\"></nz-avatar></ng-container>\n        </div>\n        <div *ngIf=\"title || desc\" class=\"ant-card-meta-detail\">\n            <div *ngIf=\"title\" class=\"ant-card-meta-title\">\n                <ng-container *ngIf=\"_title; else _titleTpl\">{{ _title }}</ng-container>\n            </div>\n            <div *ngIf=\"desc\" class=\"ant-card-meta-description\">\n                <ng-container *ngIf=\"_desc; else _descTpl\">{{ _desc }}</ng-container>\n            </div>\n        </div>\n    </div>\n    "
                },] },
    ];
    /** @nocollapse */
    NzCardMetaComponent.ctorParameters = function () { return []; };
    NzCardMetaComponent.propDecorators = {
        "_nzListItemMeta": [{ type: HostBinding, args: ['class.ant-card-meta',] },],
        "nzAvatar": [{ type: Input },],
        "nzTitle": [{ type: Input },],
        "nzDescription": [{ type: Input },],
    };
    return NzCardMetaComponent;
}());
export { NzCardMetaComponent };
