// @delon/acl/directives/acl.directive.js
import { Directive, Input, ElementRef, Renderer2 } from '@angular/core';
import { ACLService } from '../services/acl.service';
var ACLDirective = /** @class */ (function () {
    function ACLDirective(el, renderer, srv) {
        this.el = el;
        this.renderer = renderer;
        this.srv = srv;
    }
    Object.defineProperty(ACLDirective.prototype, "acl", {
        set: function (value) {
            this.set(value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ACLDirective.prototype, "ability", {
        set: function (ability) {
            this.set({
                ability: [ability],
                role: null
            });
        },
        enumerable: true,
        configurable: true
    });
    ACLDirective.prototype.set = function (value) {
        var CLS = 'acl__hide';
        var el = this.el.nativeElement;
        if (this.srv.can(value)) {
            this.renderer.removeClass(el, CLS);
        }
        else {
            this.renderer.addClass(el, CLS);
        }
    };
    ACLDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[acl]'
                },] },
    ];
    /** @nocollapse */
    ACLDirective.ctorParameters = function () { return [
        { type: ElementRef, },
        { type: Renderer2, },
        { type: ACLService, },
    ]; };
    ACLDirective.propDecorators = {
        "acl": [{ type: Input, args: ['acl',] },],
        "ability": [{ type: Input, args: ['acl-ability',] },],
    };
    return ACLDirective;
}());
export { ACLDirective };
