// @delon/abc/down-file/down-file.directive.js
import { Directive, ElementRef, Input, HostListener, EventEmitter, Output } from '@angular/core';
import { saveAs } from 'file-saver';
import { _HttpClient } from '@delon/theme';
/**
 * 文件下载
 *
 * ```html
 * <button nz-button down-file http-url="assets/demo{{i}}" file-name="demo中文">{{i}}</button>
 * ```
 */
var DownFileDirective = /** @class */ (function () {
    function DownFileDirective(el, _http) {
        this.el = el;
        this._http = _http;
        /** 请求类型 */
        this.httpMethod = 'get';
        /** 成功回调 */
        this.success = new EventEmitter();
        /** 错误回调 */
        this.error = new EventEmitter();
    }
    DownFileDirective.prototype._click = function () {
        var _this = this;
        this.el.nativeElement.disabled = true;
        this._http.request(this.httpMethod, this.httpUrl, {
            params: this._http.parseParams(this.httpData || {}),
            responseType: 'blob',
            observe: 'response'
        }).subscribe(function (res) {
            if (res.body.size <= 0) {
                _this.error.emit(res);
                return;
            }
            var fileName = _this.fileName || res.headers.get('filename') || res.headers.get('x-filename');
            saveAs(res.body, decodeURI(fileName));
            _this.success.emit(res);
            _this.el.nativeElement.disabled = false;
        }, function (err) {
            _this.error.emit(err);
            _this.el.nativeElement.disabled = false;
        });
    };
    DownFileDirective.decorators = [
        { type: Directive, args: [{ selector: '[down-file]' },] },
    ];
    /** @nocollapse */
    DownFileDirective.ctorParameters = function () { return [
        { type: ElementRef, },
        { type: _HttpClient, },
    ]; };
    DownFileDirective.propDecorators = {
        "httpData": [{ type: Input, args: ['http-data',] },],
        "httpMethod": [{ type: Input, args: ['http-method',] },],
        "httpUrl": [{ type: Input, args: ['http-url',] },],
        "fileName": [{ type: Input, args: ['file-name',] },],
        "success": [{ type: Output },],
        "error": [{ type: Output },],
        "_click": [{ type: HostListener, args: ['click',] },],
    };
    return DownFileDirective;
}());
export { DownFileDirective };
