// ngx-tinymce/index.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { TinymceOptions } from './src/tinymce.options';
import { TinymceComponent } from './src/tinymce.component';
import { ScriptService } from './src/script.service';
// region: export
export * from './src/tinymce.options';
export { TinymceComponent } from './src/tinymce.component';
// endregion
var NgxTinymceModule = (function () {
    function NgxTinymceModule() {
    }
    NgxTinymceModule.forRoot = function (options) {
        return {
            ngModule: NgxTinymceModule,
            providers: [
                ScriptService,
                { provide: TinymceOptions, useValue: options }
            ]
        };
    };
    NgxTinymceModule.decorators = [
        { type: NgModule, args: [{
                    imports: [CommonModule],
                    declarations: [TinymceComponent],
                    exports: [TinymceComponent]
                },] },
    ];
    /** @nocollapse */
    NgxTinymceModule.ctorParameters = function () { return []; };
    return NgxTinymceModule;
}());
export { NgxTinymceModule };
