// @delon/acl/index.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { DACL_OPTIONS_TOKEN, DACL_USER_OPTIONS_TOKEN, DEFAULT } from './acl.options';
import { ACLGuard } from './services/acl-guard';
// region: import
import { ACLService } from './services/acl.service';
var SERVICES = [ACLService, ACLGuard];
// components
import { ACLDirective } from './directives/acl.directive';
var COMPONENTS = [ACLDirective];
// pipes
var PIPES = [];
// endregion
// region: export
export { ACLService } from './services/acl.service';
export { ACLDirective } from './directives/acl.directive';
export * from './acl.options';
export * from './services/acl-guard';
// components
// endregion
export function optionsFactory(options) {
    return Object.assign(DEFAULT, options);
}
var AlainACLModule = /** @class */ (function () {
    function AlainACLModule() {
    }
    AlainACLModule.forRoot = function (options) {
        return {
            ngModule: AlainACLModule,
            providers: [
                { provide: DACL_USER_OPTIONS_TOKEN, useValue: options },
                { provide: DACL_OPTIONS_TOKEN, useFactory: optionsFactory, deps: [DACL_USER_OPTIONS_TOKEN] }
            ].concat(SERVICES)
        };
    };
    AlainACLModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule
                    ],
                    declarations: COMPONENTS.concat(PIPES),
                    exports: COMPONENTS.concat(PIPES)
                },] },
    ];
    /** @nocollapse */
    AlainACLModule.ctorParameters = function () { return []; };
    return AlainACLModule;
}());
export { AlainACLModule };
