// abp-ng2-module/dist/src/message/message.service.js
import { Injectable } from '@angular/core';
var MessageService = (function () {
    function MessageService() {
    }
    MessageService.prototype.info = function (message, title) {
        return abp.message.info(message, title);
    };
    MessageService.prototype.success = function (message, title) {
        return abp.message.success(message, title);
    };
    MessageService.prototype.warn = function (message, title) {
        return abp.message.warn(message, title);
    };
    MessageService.prototype.error = function (message, title) {
        return abp.message.error(message, title);
    };
    MessageService.prototype.confirm = function (message, titleOrCallBack, callback) {
        if (typeof titleOrCallBack == 'string') {
            return abp.message.confirm(message, titleOrCallBack, callback);
        }
        else {
            return abp.message.confirm(message, titleOrCallBack);
        }
    };
    MessageService.decorators = [
        { type: Injectable },
    ];
    /** @nocollapse */
    MessageService.ctorParameters = function () { return []; };
    return MessageService;
}());
export { MessageService };
