// @delon/cache/src/cache.service.js
import { Injectable, Inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import * as moment from 'moment';
import { Observable } from 'rxjs/Observable';
import { tap, map } from 'rxjs/operators';
import { DC_STORE_STORAGE_TOKEN } from './interface';
import { DC_OPTIONS_TOKEN } from '../cache.options';
var CacheService = /** @class */ (function () {
    function CacheService(options, store, http) {
        this.options = options;
        this.store = store;
        this.http = http;
        this.memory = new Map();
        this.meta = new Set();
        this.loadMeta();
    }
    CacheService.prototype.deepGet = function (obj, path, defaultValue) {
        if (!obj)
            return defaultValue;
        if (path.length <= 1) {
            var checkObj = path.length ? obj[path[0]] : obj;
            return typeof checkObj === 'undefined' ? defaultValue : checkObj;
        }
        return path.reduce(function (o, k) { return (o || {})[k]; }, obj) || defaultValue;
    };
    // region: meta
    // region: meta
    CacheService.prototype.pushMeta = 
    // region: meta
    function (key) {
        if (this.meta.has(key))
            return;
        this.meta.add(key);
        this.saveMeta();
    };
    CacheService.prototype.removeMeta = function (key) {
        if (!this.meta.has(key))
            return;
        this.meta.delete(key);
        this.saveMeta();
    };
    CacheService.prototype.loadMeta = function () {
        var _this = this;
        var ret = this.store.get(this.options.meta_key);
        if (ret && ret.v) {
            ret.v.forEach(function (key) { return _this.meta.add(key); });
        }
    };
    CacheService.prototype.saveMeta = function () {
        var metaData = [];
        this.meta.forEach(function (key) { return metaData.push(key); });
        this.store.set(this.options.meta_key, { v: metaData, e: 0 });
    };
    /**
     * 缓存对象
     */
    /**
         * 缓存对象
         */
    CacheService.prototype.set = /**
         * 缓存对象
         */
    function (key, data, options) {
        var _this = this;
        if (options === void 0) { options = {}; }
        // expire
        var e = 0;
        if (options.expire && options.expire.length > 1) {
            var vn = options.expire.split(/(\d+)/).filter(Boolean);
            e = moment().add(+vn[0], vn[1]).unix();
        }
        if (!(data instanceof Observable)) {
            this.save(options.type, key, { v: data, e: e });
            return;
        }
        return data.pipe(tap(function (v) {
            _this.save(options.type, key, { v: v, e: e });
        }));
    };
    CacheService.prototype.save = function (type, key, value) {
        if (type === 'm') {
            this.memory.set(key, value);
        }
        else {
            this.store.set(this.options.prefix + key, value);
            this.pushMeta(key);
        }
    };
    CacheService.prototype.get = function (key, options) {
        var _this = this;
        if (options === void 0) { options = {}; }
        var value = this.memory.has(key) ? this.memory.get(key) : this.store.get(this.options.prefix + key);
        if (!value || (value.e && value.e > 0 && value.e < moment().unix())) {
            if (options.mode !== 'none' && this.options.mode === 'promise') {
                return this.http.get(key).pipe(map(function (ret) { return _this.deepGet(ret, _this.options.reName, null); }), tap(function (v) { return _this.set(key, v); }));
            }
            return null;
        }
        return value.v;
    };
    /** 获取缓存数据，若 `key` 不存在或已过期则返回 null */
    /** 获取缓存数据，若 `key` 不存在或已过期则返回 null */
    CacheService.prototype.getNone = /** 获取缓存数据，若 `key` 不存在或已过期则返回 null */
    function (key) {
        return this.get(key, { mode: 'none' });
    };
    /**
     * 获取缓存，若不存在则设置缓存对象
     */
    /**
         * 获取缓存，若不存在则设置缓存对象
         */
    CacheService.prototype.tryGet = /**
         * 获取缓存，若不存在则设置缓存对象
         */
    function (key, data, options) {
        if (options === void 0) { options = {}; }
        var ret = this.getNone(key);
        if (ret === null) {
            if (!(data instanceof Observable)) {
                this.set(key, data, options);
                return data;
            }
            return this.set(key, data, options);
        }
        return ret;
    };
    // endregion
    // region: has
    /** 是否缓存 `key` */
    // endregion
    // region: has
    /** 是否缓存 `key` */
    CacheService.prototype.has = 
    // endregion
    // region: has
    /** 是否缓存 `key` */
    function (key) {
        return this.memory.has(key) || this.meta.has(key);
    };
    // endregion
    // region: remove
    /** 移除缓存 */
    // endregion
    // region: remove
    /** 移除缓存 */
    CacheService.prototype.remove = 
    // endregion
    // region: remove
    /** 移除缓存 */
    function (key) {
        if (this.memory.has(key)) {
            this.memory.delete(key);
            return;
        }
        this.store.remove(this.options.prefix + key);
        this.removeMeta(key);
    };
    /** 清空所有缓存 */
    /** 清空所有缓存 */
    CacheService.prototype.clear = /** 清空所有缓存 */
    function () {
        var _this = this;
        this.memory.clear();
        this.meta.forEach(function (key) { return _this.store.remove(_this.options.prefix + key); });
    };
    // endregion
    // endregion
    CacheService.prototype.ngOnDestroy = 
    // endregion
    function () {
        this.memory.clear();
    };
    CacheService.decorators = [
        { type: Injectable },
    ];
    /** @nocollapse */
    CacheService.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: Inject, args: [DC_OPTIONS_TOKEN,] },] },
        { type: undefined, decorators: [{ type: Inject, args: [DC_STORE_STORAGE_TOKEN,] },] },
        { type: HttpClient, },
    ]; };
    return CacheService;
}());
export { CacheService };
