// ng-clipboard-antd/index.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { NzClipboardDirective } from './src/directive';
import { NzClipboardService } from './src/service';
import { NzClipboardConfig } from './src/config';
import { NzMessageModule, NzMessageService } from 'ng-zorro-antd';
// region: export
export { NzClipboardConfig } from './src/config';
export { NzClipboardDirective } from './src/directive';
export { NzClipboardService } from './src/service';
// endregion
var NzClipboardModule = (function () {
    function NzClipboardModule() {
    }
    NzClipboardModule.forRoot = function (config) {
        return {
            ngModule: NzClipboardModule,
            providers: [
                NzClipboardService,
                NzMessageService,
                { provide: NzClipboardConfig, useValue: config }
            ]
        };
    };
    NzClipboardModule.decorators = [
        { type: NgModule, args: [{
                    imports: [CommonModule, NzMessageModule],
                    declarations: [NzClipboardDirective],
                    exports: [NzClipboardDirective]
                },] },
    ];
    /** @nocollapse */
    NzClipboardModule.ctorParameters = function () { return []; };
    return NzClipboardModule;
}());
export { NzClipboardModule };
